symless/synergy
# Wayland support (port Red Hat `libei` and `libportal` impl)

Special thanks to Peter Hutterer (@whot) and Olivier Fourdan (@ofourdan) from Red Hat for their work on implementing Wayland support.

The plan for now is to use the old events system (`REGISTER_EVENT` to create functions for event types), and in a future PR we will use the new event system by Povilas Kanapickas (@p12tic). Writing C++ `new` and `malloc` code in 2024 hurts.

https://symless.atlassian.net/browse/S1-1813
https://symless.atlassian.net/browse/S1-1812
https://symless.atlassian.net/browse/S1-1816

Wayland issues: #7420, #4090, #7256, #7184, #7127, #7125, #7105, #6690, #6610, #6558, #6196, #5803, #5802, #5763, #5747

---

**Test instructions:**
Note: GNOME 46 is required: `gnome-shell --version`

1. Follow the compile instructions in the [README](https://github.com/symless/synergy/blob/master/README.md)
2. *Replace your system-installed `libportal` with the latest version of `libportal` from git:
   1. Remove `libportal-dev` or `libportal-devel` package
   1. `build/python/bin/meson setup build/meson -Dsystem_libportal=false --reconfigure`
   2. `sudo build/python/bin/meson install -C build/meson`
   3. Verify that the new `libportal.so` was copied to system libs
   4. Rename the system-installed `libportal.so.1.0.0` to `libportal.so.1.0.0_old`
5. Run `build/bin/synergy` on Linux client and server

*Necessary until `libportal` release a new version after 0.7.1.

---

**[TODO]** Raise new issues:
- Maybe default to X11 if `XDG_SESSION_TYPE` env is not `"wayland"`?
- Maybe fall back to X11 if not GNOME46?
- Use enum based events system (based on work by Povilas Kanapickas)
- Undo all of the `new` and `mallocs` introduced in this PR (use shared/unique pointers)
- Remember annoying permission dialog states (asked every time the app starts/resumes)
- Try [`pipe2(pipefd, O_NONBLOCK)`](https://github.com/symless/synergy/pull/7449#discussion_r1738307300) in `src/lib/platform/EiEventQueueBuffer.cpp`
- Bug: System tray not working on Wayland (endless retry loop)
- Bug: Prevent computer from sleeping no longer works on Wayland
- Bug: Copy and paste doesn't seem to work on Wayland
- Bug: Mouse is not hidden on Wayland server or client when leaving screen
